uefi-logger/src/lib.rs

```diff
@@ -46,11 +47,10 @@ impl log::Log for Logger {
     }
 
     fn log(&self, record: &log::Record) {
-        let args = record.args();
-
         let writer = unsafe { &mut *self.writer.get() };
-        use core::fmt::Write;
-        writeln!(writer, "{}", args).unwrap();
+        DecoratedLog::write(writer,
```

Comment:
Instead of adding a `DecoratedLog` indirection, I think we should keep things simple and just modify the `writeln` call to write the correct format, something similar to [what the log documentation suggests](https://docs.rs/log/0.4.5/log/#implementing-a-logger):

```rust
fn log(&self, record: &Record) {
    if !self.enabled(record.metadata()) {
        return false;
    }
    writeln!(output, "{}: {}", record.level(), record.args());
}
```

Reply:
I initially did this, but then I thought that the user may print logs that span multiple lines, either voluntarily or involuntarily (some times have a multi-line Display or Debug implementation, for example the `#?` format of structs is very convenient).

When that happens, making sure that the log level is injected at the beginning at every line and not just the first has two advantages:

- It looks better, as the relative indentation of the original log output is strictly preserved
- It is easier to parse on the host side, as you only need to look at the beginning of a line of output to tell what kind of output it is (no need for log terminators and supplementary parser state).

A third option is to allow allocation in uefi-logger, in which case we can just print the log to a String and iterate on the lines of that string after the fact. That was my second attempt, but at the time I was not too satisfied about introducing allocation requirements and an extra copy just for this feature, which is how I ended up with this slightly more complex design.

Reply:
Ok, I guess we'll use with this for now. I don't think extra allocations would be a good idea in this code